## Grid World Example


For this example, we have a 2x2 grid, thus four states for each square in the grid. In the grid, you can move down, up, left and right, thus giving us our four actions. We then have a whole list for the probabilities of landing in state s' from state s, given you have applied action a. Each movement is assosciated with a reward. Once you reach the bottom right corner, there is no way to leave that state.

### Information

States = { TL = Top Left, TR = Top Right, BL = Bottom Left, BR = Bottom Right }

Actions = { U = Up, R = Right, D = Down, L = Left }

Probabilities + Rewards:

| s  | a | s’ | P(s’\|s,a) | R(s’\|s,a) |
|:--:|:--:|:--:|:----------:|:----------:|
| TL | R | TR | 0.9        | -1         |
| TL | R | BL | 0.1        | -2         |
| TL | D | BL | 0.9        | -2         |
| TL | D | TR | 0.1        | -1         |
| TR | L | TL | 0.9        | -3/2       |
| TR | L | BR | 0.1        | 10         |
| TR | D | BR | 0.8        | 15         |
| TR | D | TL | 0.2        | -1         |
| BL | R | BR | 0.9        | 20         |
| BL | R | TL | 0.1        | -5/2       |
| BL | U | TL | 0.8        | -1/2       |
| BL | U | BR | 0.2        | 5          |

To be eligible for our function, we need to present this information as two lists and two functions, which is shown below:

In [11]:
S2=["TL","TR","BL","BR"]
A2=["U","R","D","L"]
def P2(sdash,s,a):
    if s == "BR":
        return 0
    elif s == "TL":
        if a == "R":
            if sdash == "TR":
                return 0.9
            elif sdash == "BL":
                return 0.1
        elif a == "D":
            if sdash == "BL":
                return 0.9
            elif sdash == "TR":
                return 0.1
    elif s == "TR":
        if a == "L":
            if sdash == "TL":
                return 0.9
            elif sdash == "BR":
                return 0.1
        elif a == "D":
            if sdash == "BR":
                return 0.8
            elif sdash == "TL":
                return 0.2
    elif s == "BL":
        if a == "R":
            if sdash == "BR":
                return 0.9
            elif sdash == "TL":
                return 0.1
        elif a == "U":
            if sdash == "TL":
                return 0.8
            elif sdash == "BR":
                return 0.2
    return 0
    
def R2(sdash,s,a):
    if s == "BR":
        return 0
    elif s == "TL":
        if a == "R":
            if sdash == "TR":
                return -1
            elif sdash == "BL":
                return -2
        elif a == "D":
            if sdash == "BL":
                return -2
            elif sdash == "TR":
                return -1
    elif s == "TR":
        if a == "L":
            if sdash == "TL":
                return -3/2
            elif sdash == "BR":
                return 10
        elif a == "D":
            if sdash == "BR":
                return 15
            elif sdash == "TL":
                return -1
    elif s == "BL":
        if a == "R":
            if sdash == "BR":
                return 20
            elif sdash == "TL":
                return -5/2
        elif a == "U":
            if sdash == "TL":
                return -1/2
            elif sdash == "BR":
                return 5
    return 0

Since our reward function takes s', we want to use mode 2 of our function. We then pick the usual $\gamma=0.9$ and set $\epsilon=0.00001$, as we want to have a really low epsilon to make sure we have converged to the correct values. We can now import our Value_iteration function from our package and run it.

In [12]:
from ValueIt.ValueItFunction import Value_iteration 

Value_iteration(S2,A2,P2,R2,0.9,0.00001,2)

[['D', 'D', 'R', 'U'],
 [14.86387023427788, 14.475495422783593, 19.087747711391795, 0.0]]

Each states optimal policy and associated value can be found in the table below. The only slight caveat is that our code will return an optimal policy for BR, even though we know you can not leave this state, which is why the value is zero. We unfortunately just have to ignore this.

| State | Optimal Policy | Value |
|:-----:|:--------------:|:-----:|
| TL    | D              | 14.864|
| TR    | D              | 14.475|
| BL    | R              | 19.088|
| BR    |            N/A | 0.000 |